### PR TITLE
VISSL initial integration

### PR DESCRIPTION
**Pull request by ananyahjha93**
## What does this PR do?

POC PR for VISSL

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/integrations/vissl/adapter.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+if _VISSL_AVAILABLE:
+    from classy_vision.losses import build_loss, ClassyLoss
+    from classy_vision.tasks import ClassyTask
+    from flash.core.integrations.vissl.base_ssl_model import BaseSSLMultiInputOutputModel
+
+    from flash.core.integrations.vissl.hooks import AdaptVISSLHooks
+
+
+class MockVISSLTask:
```
Comment (tchaton) [inline flash/core/integrations/vissl/adapter.py:42]: Mind adding comments, docstring ?

File: flash/core/integrations/vissl/adapter.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    def __init__(self, vissl_loss, task_config, vissl_model) -> None:
+        self.loss = vissl_loss
+        self.config = task_config
+        self.model = vissl_model
+
+        self.device = torch.device('cpu')
```
Comment (tchaton) [inline flash/core/integrations/vissl/adapter.py:48]: Why do you hard-code the device ?

File: flash/core/integrations/vissl/adapter.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+        self.iteration = 0
+        self.max_iteration = 100000 # set this correctly
```
Comment (tchaton) [inline flash/core/integrations/vissl/adapter.py:51]: Mind sharing more details on those values ?

File: flash/core/integrations/vissl/adapter.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        return cls(
+            backbone=backbone,
+            head=head,
+            loss_fn=loss_fn,
+            embedding_dim=embedding_dim,
+            hooks=hooks,
+            **kwargs,
+        )
+
+    def _generate_model_config(self):
```
Comment (tchaton) [inline flash/core/integrations/vissl/adapter.py:182]: Can we use the config from VISSL directly ? This won't work for other models.

File: flash/core/integrations/vissl/adapter.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    def training_step(self, batch: Any, batch_idx: int) -> Any:
+        out = self(batch[DefaultDataKeys.INPUT])
+        self.task.last_batch['sample']['input'] = batch[DefaultDataKeys.INPUT]
+
+        # call forward hook from VISSL (momentum updates)
+        for hook in self.hooks:
+            hook.on_forward(self.vissl_task)
+
+        # out can be torch.Tensor/List target is torch.Tensor
+        # loss = self.vissl_loss(out, target)
```
Comment (tchaton) [inline flash/core/integrations/vissl/adapter.py:255]: Where is the loss returned ? Where is the optimization performed ?

File: flash/core/integrations/vissl/base_ssl_model.py [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+from vissl.utils.checkpoint import (
+    CheckpointLoader,
+    init_model_from_consolidated_weights,
+)
+from vissl.utils.env import get_machine_local_and_dist_rank
+from vissl.utils.fsdp_utils import fsdp_recursive_reset_lazy_init
+from vissl.utils.misc import set_torch_seed
+
+
+class BaseSSLMultiInputOutputModel(ClassyModel):
```
Comment (tchaton) [inline flash/core/integrations/vissl/base_ssl_model.py:32]: This class seems to contain lot of boilerplate. Is it fully required ?

File: flash/core/integrations/vissl/base_ssl_model.py [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+            key = self.model_config.MULTI_INPUT_HEAD_MAPPING[input_idx][0]
+            eval_layer_names = self.model_config.MULTI_INPUT_HEAD_MAPPING[input_idx][1]
+            if key in self._input_to_eval_features_map:
+                raise ValueError(
+                    f"duplicate key {key} \
+                    specified for MODEL.MULTI_INPUT_HEAD_MAPPING."
+                )
+            self._input_to_eval_features_map[key] = eval_layer_names
+
+    def get_classy_state(self, deep_copy=False):
```
Comment (tchaton) [inline flash/core/integrations/vissl/base_ssl_model.py:393]: I don't think the model should contain this. Lightning should be responsible to handle the checkpoiting.

File: flash/core/integrations/vissl/dino_hook.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+def build_model(model_config, optimizer_config):
+    """
+    Given the model config and the optimizer config, construct the model.
+    The returned model is not copied to gpu yet (if using gpu) and neither
+    wrapped with DDP yet. This is done later train_task.py .prepare()
+    """
+    return BaseSSLMultiInputOutputModel(model_config, optimizer_config)
+
+
+class DINOHook(ClassyHook):
```
Comment (tchaton) [inline flash/core/integrations/vissl/dino_hook.py:30]: Do you need this ? Can't be imported from VISSL ?

**Review by ethanwharris — APPROVED** [review]
LGTM! Docs, docstrings etc. to come in follow-up :smiley:

**Comment by ananyahjha93** [comment]
@ethanwharris The image classification notebook is failing when loading the checkpoint from s3.

File: flash/core/registry.py [hunk 9]
```diff
@@ -111,7 +111,11 @@ def _register_function(
         if not callable(fn):
             raise MisconfigurationException(f"You can only register a callable, found: {fn}")
 
-        name = name or fn.__name__
+        if name is None:
+            if hasattr(fn, "func"):
+                name = fn.func.__name__
+            else:
+                name = fn.__name__
```
Comment (tchaton) [inline flash/core/registry.py:118]:
```suggestion
        if name is None:
            name = fn.func.__name__ if hasattr(fn, "func") else fn.__name__

```

File: flash/image/embedding/backbones/vissl_backbones.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    num_heads: int = 6,
+    mlp_dim: int = 1532,
+    dropout_rate: float = 0,
+    attention_dropout_rate: float = 0,
+    drop_path_rate: float = 0,
+    qkv_bias: bool = True,
+    qk_scale: bool = False,
+    classifier: str = "token",
+    **kwargs,
+) -> nn.Module:
```
Comment (tchaton) [inline flash/image/embedding/backbones/vissl_backbones.py:43]: some docstrings would be good. Same below.

File: flash/image/embedding/heads/vissl_heads.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+if _VISSL_AVAILABLE:
+    from vissl.config.attr_dict import AttrDict
+    from vissl.models.heads import MODEL_HEADS_REGISTRY, register_model_head
+
+    from flash.image.embedding.vissl.adapter import VISSLAdapter
+else:
+    AttrDict = object
+
+
+class SimCLRHead(nn.Module):
```
Comment (tchaton) [inline flash/image/embedding/heads/vissl_heads.py:31]: Not sure this is fully required, but I might be wrong.

File: flash/image/embedding/vissl/adapter.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        hooks: List[ClassyHook],
+    ) -> Adapter:
+        result = cls(
+            backbone=backbone,
+            head=head,
+            loss_fn=loss_fn,
+            hooks=hooks,
+        )
+
+        result.__dict__["adapter_task"] = task
```
Comment (tchaton) [inline flash/image/embedding/vissl/adapter.py:116]: Neat !

File: flash/image/embedding/vissl/adapter.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+        self.adapter_task.log_dict({"val_loss": loss})
+
+        return loss
+
+    def test_step(self, batch: Any, batch_idx: int) -> None:
+        out = self.ssl_forward(batch[DefaultDataKeys.INPUT])
+        self.task.last_batch["sample"]["input"] = batch[DefaultDataKeys.INPUT]
+
+        loss = self.loss_fn(out, target=None)
+        self.adapter_task.log_dict({"test_loss": loss})
```
Comment (tchaton) [inline flash/image/embedding/vissl/adapter.py:193]: Same here.

File: flash/image/embedding/vissl/adapter.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+        return loss
+
+    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Any:
```
Comment (tchaton) [inline flash/image/embedding/vissl/adapter.py:197]: Did you test this one ?

File: flash/image/embedding/vissl/hooks.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+    def __init__(self):
+        super().__init__()
+
+    @torch.no_grad()
+    def on_start(self, task: "flash.image.embedding.vissl.adapter.MockVISSLTask") -> None:
+        lightning_module = task.vissl_adapter.adapter_task
+        task.device = lightning_module.device
+
+        num_nodes = lightning_module.trainer.num_nodes
```
Comment (tchaton) [inline flash/image/embedding/vissl/hooks.py:49]: Some comments there would be nice ! Nice :)

File: flash_examples/image_embedder.py [hunk 16]
```diff
@@ -11,15 +11,51 @@
...
+    batch_size=16,
+)
 
 # 2. Build the task
-embedder = ImageEmbedder(backbone="resnet101")
+embedder = ImageEmbedder(
+    backbone="resnet",
+    training_strategy="barlow_twins",
+    head="simclr_head",
+    latent_embedding_dim=128,
```
Comment (tchaton) [inline flash_examples/image_embedder.py:47]: Let's adopt the training_stategies_kwargs={...} API for now.

**Review by tchaton — APPROVED** [review]
Great work ! Definitely looks much better now :) Still some comment, but can be addressed in the following PR.

File: flash/image/embedding/heads/vissl_heads.py (hunk 11 above)
Comment (ananyahjha93) [reply flash/image/embedding/heads/vissl_heads.py:31]: VISSL creates MLP heads for SimCLR/Barlow Twins/MOCO?barlow Twins etc. layer by layer, in a yaml config, which is not practical for flash.
